JLio.Commands/Advanced/Compare.cs

```diff
@@ -0,0 +1,482 @@
+                result.IsValid = false;
+            }
+
+            if (Settings == null)
+            {
+                result.ValidationMessages.Add($"Settings property for {CommandName} command is missing");
+                result.IsValid = false;
+            }
+
+            return result;
+        }
+
+        public override JLioExecutionResult Execute(JToken dataContext, IExecutionOptions executionOptions)
+        {
+            options = executionOptions;
+            var compareResults = new CompareResults();
+            var validationResult = ValidateCommandInstance();
+            if (!validationResult.IsValid)
+            {
+                validationResult.ValidationMessages.ForEach(i =>
+                    options.Logger?.Log(LogLevel.Warning, Constants.CommandExecution, i));
+                return new JLioExecutionResult(false, dataContext);
+            }
+
+            foreach (var source in options.ItemsFetcher.SelectTokens(FirstPath, dataContext))
+            foreach (var target in options.ItemsFetcher.SelectTokens(SecondPath, dataContext))
+                compareResults.AddRange(CompareTokens(source, target));
+            var filteredResults = FilterCompareResult(compareResults);
+            SetResults(dataContext, JToken.FromObject(filteredResults));
+
+            return new JLioExecutionResult(true, dataContext);
+        }
+
+        private CompareResults FilterCompareResult(CompareResults results)
+        {
+            if (Settings.ResultTypes == null || !Settings.ResultTypes.Any()) return results;
+
+            return new CompareResults(results.Where(r => Settings.ResultTypes.Contains(r.DifferenceType)).ToList());
+        }
+
+        private void SetResults(JToken dataContext, JToken compareResults)
+        {
+            var resultTargets = dataContext.SelectTokens(ResultPath);
+
+            var enumerable = resultTargets as JToken[] ?? resultTargets.ToArray();
+
+            var script = new JLioScript();
+            if (!enumerable.Any())
+            {
+                script.Add(compareResults).OnPath(ResultPath).Execute(dataContext, options);
+                return;
+            }
+
+            foreach (var resultTarget in enumerable)
+                script.Set(compareResults).OnPath(resultTarget.Path).Execute(dataContext, options);
+        }
+
+        private CompareResults CompareTokens(JToken source, JToken target)
+        {
+            var result = new CompareResults();
+            if (source.Type != target.Type)
+            {
+                result.Add(GetDifferentTypeResult(source, target));
+            }
+            else
+            {
+                result.Add(GetSameTypeResult(source, target));
+                result.AddRange(CompareSameTypeObjects(source, target));
+            }
+
+            return result;
+        }
+
+        private CompareResults CompareSameTypeObjects(JToken source, JToken target)
```

Comment:
in my oppinion, instead of having all of these switch cases and private methods to handle each possibility, a better option to stick to the single responsibility and open-closed principles is to isolate each comparer into it's own dedicated class,   and have them available in the Compare class by injecting a dictionary of these comparers (key should be the JTokenType).
Then, instead of doing a switch statement, you pick by key from the dictionary and select the proper comparer for that particular type of token.
This enables extending the behavior by including a different item in the dictionary, for a different type of token, without having to come and change the Compare.cs file (to add new private methods and new switch statements)

Reply:
there will not be any new types introduced since they are JSON types. It will clutter the project heavily with classes that hold only 1 method.

Reply:
if it is needed. add a item in the quality project so we can fix this later during a refactor session